src/MongoDB.Driver.Core/Core/Servers/RoundTripTimeMonitor.cs

```diff
@@ -0,0 +1,164 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Diagnostics;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Servers
+{
+    internal interface IRoundTripTimeMonitor : IDisposable
+    {
+        TimeSpan Average { get; }
+        void AddSample(TimeSpan roundTripTime);
+        void Reset();
+        Task RunAsync();
+    }
+
+    internal class RoundTripTimeMonitor : IRoundTripTimeMonitor
+    {
+        private readonly ExponentiallyWeightedMovingAverage _averageRoundTripTimeCalculator = new ExponentiallyWeightedMovingAverage(0.2);
+        private readonly CancellationToken _cancellationToken;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly IConnectionFactory _connectionFactory;
+        private readonly EndPoint _endPoint;
+        private readonly TimeSpan _heartbeatInterval;
+        private readonly object _lock = new object();
+        private IConnection _roundTripTimeConnection;
+        private readonly ServerId _serverId;
+
+        public RoundTripTimeMonitor(
+            IConnectionFactory connectionFactory,
+            ServerId serverId,
+            EndPoint endpoint,
+            TimeSpan heartbeatInterval)
+        {
+            _connectionFactory = Ensure.IsNotNull(connectionFactory, nameof(connectionFactory));
+            _serverId = Ensure.IsNotNull(serverId, nameof(serverId));
+            _endPoint = Ensure.IsNotNull(endpoint, nameof(endpoint));
+            _heartbeatInterval = heartbeatInterval;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+        }
+
+        public TimeSpan Average
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _averageRoundTripTimeCalculator.Average;
+                }
+            }
+        }
+
+        // public methods
+        public void Dispose()
+        {
+            _cancellationTokenSource.Cancel();
+            _cancellationTokenSource.Dispose();
+
+            try { _roundTripTimeConnection?.Dispose(); } catch { }
+        }
```

Comment:
The usual implementation of Dispose uses a flag to prevent a second call to Dispose from doing anything:

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            _cancellationTokenSource.Cancel();
            _cancellationTokenSource.Dispose();

            try { _roundTripTimeConnection?.Dispose(); } catch { }
        }
    }

Reply:
Also... if you don't intend to provide a `Disposed(bool disposing)` overload the class should be `sealed`.

Reply:
done